### PR TITLE
fix(config): Add missing handling for `envCache` in `getKeys()`

### LINKED ISSUE #36126: Overridden `config.php` variables (via `NC_*` are not reflected in `occ config:system:get` command
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] Nextcloud Server **is** running on 64bit capable CPU, PHP and OS.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

PR [3966](https://github.com/nextcloud/server/pull/3966) introduced the possibility to override values in `config.php` via environment variables (a feature frequently used in container setups).

However, `occ config:system:get` does not reflect these variables. This could break automation or external tools if they rely on this command to extract things like database configuration etc.

### Steps to reproduce

1. Setup a NextCloud instance and define the database via environment variables `NC_dbhost`, `NC_dbuser` etc. instead of `config.php`
2. Run `occ config:system:get dbhost`
3. Empty result

### Expected behavior

Proper config value returned

### Installation method

Community Docker image

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 8.1

### Web server

Apache (supported)

### Database engine version

MariaDB

### Is this bug present after an update or on a fresh install?

None

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
{
    "system": {
        "version": "25.0.2.3",
        "installed": true,
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "htaccess.RewriteBase": "\/",
        "dbtype": "mysql",
        "mysql.utf8mb4": true,
        "log_type": "file",
        "loglevel": 0,
        "file": "\/var\/www\/html\/data\/nc.log",
        "theme": "",
        "maintenance": false,
        "trusted_domains": [
            "***REMOVED SENSITIVE VALUE***"
        ],
        "trusted_proxies": "***REMOVED SENSITIVE VALUE***",
        "forcessl": true,
        "overwrite.cli.url": "***REMOVED SENSITIVE VALUE***",
        "overwriteprotocol": "https",
        "session_relaxed_expiry": false,
        "secret": "***REMOVED SENSITIVE VALUE***",
        "share_folder": "\/Shared",
        "trashbin_retention_obligation": "auto, 7",
        "versions_retention_obligation": "auto, 7",
        "default_phone_region": "DE",
        "ldapIgnoreNamingRules": false,
        "ldapProviderFactory": "\\OCA\\User_LDAP\\LDAPProviderFactory",
        "memcache.local": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 6379,
            "password": "***REMOVED SENSITIVE VALUE***"
        },
        "apps_paths": [
            {
                "path": "\/var\/www\/html\/apps",
                "url": "\/apps",
                "writable": false
            },
            {
                "path": "\/var\/www\/html\/custom_apps",
                "url": "\/custom_apps",
                "writable": true
            }
        ],
        "appstore.experimental.enabled": true,
        "updater.release.channel": "stable",
        "mail_smtpmode": "smtp",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "587",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_sendmailmode": "smtp",
        "mail_smtpauthtype": "PLAIN",
        "mail_smtpauth": "1",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpsecure": "tls",
        "allow_local_remote_servers": true,
        "app_install_overwrite": [
            "impersonate",
            "socialsharing_email",
            "fulltextsearch",
            "fulltextsearch_elasticsearch",
            "files_fulltextsearch",
            "files_fulltextsearch_tesseract",
            "mediadc"
        ],
        "enable_previews": true,
        "preview_imaginary_url": "***REMOVED SENSITIVE VALUE***",
        "preview_max_x": 1024,
        "preview_max_y": 1024,
        "enabledPreviewProviders": [
            "OC\\Preview\\PNG",
            "OC\\Preview\\JPEG",
            "OC\\Preview\\GIF",
            "OC\\Preview\\HEIC",
            "OC\\Preview\\BMP",
            "OC\\Preview\\XBitmap",
            "OC\\Preview\\MarkDown",
            "OC\\Preview\\MP3",
            "OC\\Preview\\TXT",
            "OC\\Preview\\HEIC",
            "OC\\Preview\\OpenDocument",
            "OC\\Preview\\Movie",
            "OC\\Preview\\MKV",
            "OC\\Preview\\MP4",
            "OC\\Preview\\AVI",
            "OC\\Preview\\Krita",
            "OC\\Preview\\Imaginary"
        ],
        "activity_expire_days": 365,
        "activity_use_cached_mountpoints": true,
        "overwritehost": "***REMOVED SENSITIVE VALUE***"
    }
}
```

### List of activated Apps

```shell
Enabled:
  - activity: 2.17.0
  - admin_audit: 1.15.0
  - audioplayer: 3.3.1
  - bookmarks: 12.0.0
  - bruteforcesettings: 2.5.0
  - calendar: 4.2.1
  - camerarawpreviews: 0.8.0
  - circles: 25.0.0
  - cloud_federation_api: 1.8.0
  - cloud_py_api: 0.1.1
  - comments: 1.15.0
  - contacts: 5.0.2
  - contactsinteraction: 1.6.0
  - cookbook: 0.10.1
  - cospend: 1.5.5
  - dashboard: 7.5.0
  - dav: 1.24.0
  - deck: 1.8.3
  - external: 5.0.0
  - facerecognition: 0.9.11
  - federatedfilesharing: 1.15.0
  - federation: 1.15.0
  - files: 1.20.1
  - files_automatedtagging: 1.15.0
  - files_downloadactivity: 1.15.0
  - files_external: 1.17.0
  - files_fulltextsearch: 25.0.0
  - files_fulltextsearch_tesseract: 25.0.0
  - files_linkeditor: 1.1.13
  - files_mindmap: 0.0.27
  - files_pdfviewer: 2.6.0
  - files_rightclick: 1.4.0
  - files_sharing: 1.17.0
  - files_trashbin: 1.15.0
  - files_versions: 1.18.0
  - firstrunwizard: 2.14.0
  - forms: 3.0.3
  - fulltextsearch: 25.0.0
  - fulltextsearch_elasticsearch: 25.0.0
  - gpxpod: 5.0.4
  - groupfolders: 13.1.0
  - guests: 2.3.0
  - health: 1.6.1
  - impersonate: 1.11.0
  - logreader: 2.10.0
  - lookup_server_connector: 1.13.0
  - mail: 2.2.2
  - maps: 0.2.4
  - mediadc: 0.3.2
  - metadata: 0.17.0
  - music: 1.8.1
  - news: 20.0.0
  - nextcloud_announcements: 1.14.0
  - notes: 4.6.0
  - notifications: 2.13.1
  - notify_push: 0.5.0
  - oauth2: 1.13.0
  - onlyoffice: 7.6.8
  - password_policy: 1.15.0
  - phonetrack: 0.7.2
  - photos: 2.0.1
  - polls: 4.0.0
  - previewgenerator: 5.1.1
  - privacy: 1.9.0
  - provisioning_api: 1.15.0
  - quota_warning: 1.15.0
  - recognize: 3.3.4
  - recommendations: 1.4.0
  - related_resources: 1.0.3
  - serverinfo: 1.15.0
  - settings: 1.7.0
  - sharebymail: 1.15.0
  - side_menu: 3.5.1
  - socialsharing_email: 2.5.0
  - spreed: 15.0.2
  - support: 1.8.0
  - survey_client: 1.13.0
  - suspicious_login: 4.3.0
  - systemtags: 1.15.0
  - tasks: 0.14.5
  - text: 3.6.0
  - theming: 2.0.1
  - twofactor_backupcodes: 1.14.0
  - twofactor_email: 2.7.1
  - twofactor_gateway: 0.20.0
  - twofactor_totp: 7.0.0
  - twofactor_webauthn: 1.0.0
  - unsplash: 2.1.1
  - updatenotification: 1.15.0
  - user_status: 1.5.0
  - viewer: 1.9.0
  - weather_status: 1.5.0
  - workflowengine: 2.7.0
Disabled:
  - encryption: 2.2.0
  - files_accesscontrol: 1.15.1
  - files_markdown: 2.3.6
  - user_ldap: 1.14.1
  - user_saml: 5.1.2
```

### Nextcloud Signing status

_No response_

### Nextcloud Logs

_No response_

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by joshtrichards**
*Context: #3966*

* Resolves: #36126  

## Summary

`NC_` environment variable `config.php` overrides were not appearing in the output of `occ config:list system` nor `occ config:system:get xxx`. This was creating nearly impossible to diagnose configuration/ behavior discrepancies.

`getValue()` was already, of course, returning `NC_` provided config values. But `getKeys()` was not due to (a) it having no handling for the `envCache`; (b) all values in `envCache` being saved under `NC_` keys rather than their actual config keys (which needed additional handling to be merged with the non-`NC_` keys).

- Refactored `readData()` so that:
  - we aren't saving the entire environment in the `envCache` anymore (only those prefixed `NC_`)
  - we save `NC_` provided config values in the `envCache` under their real configuration key
- Refactored `getValue()` to accommodate `readData()` refactor
- Fixed `getKeys()` to now return `envCache` keys as well in order to be consistent with `getValue()` (in turn, addressing the underlying bugs noted in the opening paragraph)

Result:
- Environment provided config variables now appear in `occ config:list system` as expected.
- Environment provided config variables now appear when queried via `occ config:system:get blah`
- `envCache` is now free of non-NC related environment variables

## TODO

- [x] Update/add tests (looking now)
 
## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Comment by joshtrichards** [comment]
Cypress test seems unrelated (but not the usual "just re-run it" Cypress test failure either).

File: lib/private/Config.php [hunk 1]
```diff
@@ -226,7 +225,15 @@ private function readData() {
 			}
 		}
 
-		$this->envCache = getenv();
+		// grab any "NC_" environment variables
+		$envRaw = getenv();
+		// only save environment variables prefixed with "NC_" in the cache
+		foreach ($envRaw as $rawEnvKey => $rawEnvValue) {
+			if (str_starts_with($rawEnvKey, self::ENV_PREFIX)) {
+				$realKey = explode(self::ENV_PREFIX, $rawEnvKey)[1];
```
Comment (come-nc) [inline lib/private/Config.php:233]: substr would me faster I think
```suggestion
				$realKey = substr($rawEnvKey,0,strlen(self::ENV_PREFIX));
```
Ideally with the strlen stored in a var before the foreach.

Comment (joshtrichards) [reply lib/private/Config.php]: Good call. Thanks. Adjusted your suggestion ever so slightly (I want the second half, not the prefix itself).

**Review by come-nc — APPROVED** [review]

**Review by icewind1991 — APPROVED** [review]

**Pull request merged**